mempool/cat/cache.go

```diff
@@ -0,0 +1,291 @@
+package cat
+
+import (
+	"container/list"
+	"time"
+
+	tmsync "github.com/tendermint/tendermint/libs/sync"
+	"github.com/tendermint/tendermint/types"
+)
+
+// LRUTxCache maintains a thread-safe LRU cache of raw transactions. The cache
+// only stores the hash of the raw transaction.
+// NOTE: This has been copied from mempool/cache with the main diffence of using
+// tx keys instead of raw transactions.
+type LRUTxCache struct {
+	staticSize int
+
+	mtx      tmsync.Mutex
+	cacheMap map[types.TxKey]*list.Element
+	list     *list.List
+}
+
+func NewLRUTxCache(cacheSize int) *LRUTxCache {
+	return &LRUTxCache{
+		staticSize: cacheSize,
+		cacheMap:   make(map[types.TxKey]*list.Element, cacheSize),
+		list:       list.New(),
+	}
+}
+
+func (c *LRUTxCache) Reset() {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	c.cacheMap = make(map[types.TxKey]*list.Element, c.staticSize)
+	c.list.Init()
+}
+
+func (c *LRUTxCache) Push(txKey types.TxKey) bool {
+	if c.staticSize == 0 {
+		return true
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	moved, ok := c.cacheMap[txKey]
+	if ok {
+		c.list.MoveToBack(moved)
+		return false
+	}
+
+	if c.list.Len() >= c.staticSize {
+		front := c.list.Front()
+		if front != nil {
+			frontKey := front.Value.(types.TxKey)
+			delete(c.cacheMap, frontKey)
+			c.list.Remove(front)
+		}
+	}
+
+	e := c.list.PushBack(txKey)
+	c.cacheMap[txKey] = e
+
+	return true
+}
+
+func (c *LRUTxCache) Remove(txKey types.TxKey) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	e := c.cacheMap[txKey]
+	delete(c.cacheMap, txKey)
+
+	if e != nil {
+		c.list.Remove(e)
+	}
```

Comment:
Curious to know why the `delete` operation is not moved to its following `if` block? the current version does not error out as `delete` is a no-op for the invalid and non-exiting keys, but logically it would make sense for both `Remove` and `delete` operations to be conditioned to `e` being `nil`, wdyt?